File: Languages/PHP/Day026.php
<!-- Languages - PHP - Day 26 -->
<!-- Foundational Concepts: Basics, syntax, loops, conditionals, functions, and variables. -->

<!-- Create a function that determines if a given number is a Happy Number. 


<?php

function isHappyNumber($num) {
    $seen = [];
    while ($num != 1 && !isset($seen[$num])) {
        $seen[$num] = true;
        $num = sumOfSquares($num);
    }
    return $num == 1;
}

function sumOfSquares($num) {
    $sum = 0;
    while ($num > 0) {
        $digit = $num % 10;
        $sum += $digit * $digit;
        $num = (int)($num / 10);
    }
    return $sum;
}

while (true) {
    echo "Enter a number (or 'q' to quit): ";
    $input = trim(fgets(STDIN));

    if (strtolower($input) === 'q') {
        echo "Exiting program. Goodbye!\n";
        break;
    }

    if (!filter_var($input, FILTER_VALIDATE_INT) && $input != "0") {
        echo "Invalid input. Please enter an integer.\n";
        continue;
    }

    $num = (int)$input;
    if ($num < 0) {
        echo "Negative numbers are not considered Happy Numbers.\n";
    } else {
        $result = isHappyNumber($num);
        echo "The number $num is " . ($result ? "a Happy Number." : "not a Happy Number.") . "\n";
    }
}

?>

File: Languages/Python/Day026.py
# Languages - Python - Day 26
# Foundational Concepts: Basics, syntax, loops, conditionals, functions, and variables.

# Create a function that determines if a given number is a Happy Number. 


def is_happy_number(num):
    seen = set()
    while num != 1 and num not in seen:
        seen.add(num)
        num = sum_of_squares(num)
    return num == 1


def sum_of_squares(num):
    sum_result = 0
    while num > 0:
        digit = num % 10
        sum_result += digit**2
        num //= 10
    return sum_result


def main():
    while True:
        num_input = input("Enter a number (or 'q' to quit): ").strip()

        if num_input.lower() == "q":
            print("Exiting program. Goodbye!")
            break

        try:
            num = int(num_input)
            if num < 0:
                print("Negative numbers are not considered Happy Numbers.")
            else:
                result = is_happy_number(num)
                print(
                    f"The number {num} is {'a Happy Number' if result else 'not a Happy Number'}."
                )
        except ValueError:
            print("Invalid input. Please enter an integer.")


if __name__ == "__main__":
    main()
